### PR TITLE
Utilize discriminant_value for more efficient deriving

### PR DESCRIPTION
**Pull request by Marwes**
PR for #26128.

Improves codegen in deriving by utilizing the discriminant_value intrinsic.

I have a more detailed comment on the changes in a comment on the issue [here](https://github.com/rust-lang/rust/issues/26128#issuecomment-111509863)
### Old

```
running 7 tests
test large_c_like       ... bench:   2,694,129 ns/iter (+/- 5,170)
test large_c_like_ord   ... bench:   2,723,521 ns/iter (+/- 9,098)
test test1_partial_eq   ... bench:   2,439,317 ns/iter (+/- 2,135)
test test1_partial_ord  ... bench:   2,499,114 ns/iter (+/- 55,766)
test test2_partial_eq   ... bench:   3,562,815 ns/iter (+/- 45,590)
test test2_partial_ord  ... bench:   3,398,306 ns/iter (+/- 22,180)
test test_match_success ... bench:   1,509,267 ns/iter (+/- 1,982)
```
### New

```
running 7 tests
test large_c_like       ... bench:     286,509 ns/iter (+/- 474)
test large_c_like_ord   ... bench:     286,666 ns/iter (+/- 8,756)
test test1_partial_eq   ... bench:     286,584 ns/iter (+/- 2,789)
test test1_partial_ord  ... bench:     286,470 ns/iter (+/- 516)
test test2_partial_eq   ... bench:   2,228,997 ns/iter (+/- 34,191)
test test2_partial_ord  ... bench:   1,731,699 ns/iter (+/- 21,756)
test test_match_success ... bench:   1,509,630 ns/iter (+[- 3,765)
```

[Benchmark](https://gist.github.com/Marwes/7c0b3468d0cae972a2b4)

### CODE REVIEW

**Comment by pcwalton** [comment]
Do you know if this improves compile times of rustc?

**Comment by Marwes** [comment]
@pcwalton I haven't tested the compile times so I can't say for certain. C-like enums generate a lot better IR which should help make both compiled programs faster as well as compiling that code faster. Other enums should get a less noticeable speedup as well.

I might do some benchmarking on the compiler tomorrow to see if it makes a difference though.

File: src/libsyntax/ext/deriving/generic/mod.rs [hunk 1]
```diff
@@ -1048,15 +1048,23 @@ impl<'a> MethodDef<'a> {
     /// discriminant values.  See issue #15523.)
```
Comment (bluss) [inline src/libsyntax/ext/deriving/generic/mod.rs]: I looks like this bug here was fixed, so it's a good time to update this comment too I think. I assume using discr value is always correct now?

**Comment by bluss** [comment]
Does this affect codegen for `Option<T>`? Just thinking of our most important enum.

**Comment by Marwes** [comment]
I ran the same benchmark on an Option<i32> and I got a 1.1x speedup. So I'd guess it helps a little though it might very well vary depending on what type the Option contains.

**Comment by bluss** [comment]
That's nice.

**Comment by Marwes** [comment]
@pcwalton I compiled rustc a few times with and without this PR but I got to much noise to be able to say anything.

**Comment by Marwes** [comment]
Removed the comment mentioning #15523 as per @bluss suggestion, as that had been fixed.

**Pull request merged**